python_modules/dagster/dagster/_daemon/backfill.py

```diff
@@ -1,19 +1,78 @@
+import threading
+from typing import TYPE_CHECKING, Iterable, Mapping, Optional, Set, cast
+
+import pendulum
 
 from dagster._core.execution.asset_backfill import execute_asset_backfill_iteration
 from dagster._core.execution.backfill import BulkActionStatus, PartitionBackfill
 from dagster._core.execution.job_backfill import execute_job_backfill_iteration
 from dagster._core.workspace.context import IWorkspaceProcessContext
-from dagster._utils.error import SerializableErrorInfo, serializable_error_info_from_exc_info
+from dagster._utils.error import serializable_error_info_from_exc_info
+
+if TYPE_CHECKING:
+    from .daemon import TDaemonGenerator
+
+
+ACTIVE_BACKFILLS_INTERVAL_SECONDS = 2
+
+
+def execute_backfill_iteration_loop(
+    workspace_process_context: IWorkspaceProcessContext,
+    logger: logging.Logger,
+    shutdown_event: threading.Event,
+    until=None,
+) -> "TDaemonGenerator":
+    """
+    Helper function that performs sensor evaluations on a tighter loop, while reusing grpc locations
+    within a given daemon interval.  Rather than relying on the daemon machinery to run the
+    iteration loop every 30 seconds, sensors are continuously evaluated, every 5 seconds. We rely on
+    each sensor definition's min_interval to check that sensor evaluations are spaced appropriately.
+    """
+    from .controller import DEFAULT_DAEMON_INTERVAL_SECONDS
+
+    active_backfills: Set[str] = set()
+    while True:
+        start_time = pendulum.now("UTC").timestamp()
+        if until and start_time >= until:
+            # provide a way of organically ending the loop to support test environment
+            break
+
+        for backfill in execute_backfill_iteration(workspace_process_context, logger):
+            if backfill:
+                if backfill.status.is_active:
+                    active_backfills.add(backfill.backfill_id)
+                else:
+                    active_backfills.discard(backfill.backfill_id)
+
+                yield backfill.error
+            else:
+                yield None
+
+        # Yield to check for heartbeats in case there were no yields within
+        # execute_backfill_iteration
+        yield None
+
+        end_time = pendulum.now("UTC").timestamp()
+
+        loop_duration = end_time - start_time
+        if len(active_backfills) > 0:
+            interval = ACTIVE_BACKFILLS_INTERVAL_SECONDS
+        else:
+            interval = DEFAULT_DAEMON_INTERVAL_SECONDS
+
+        sleep_time = max(0, interval - loop_duration)
+        shutdown_event.wait(sleep_time)
+
+        yield None
 
 
 def execute_backfill_iteration(
     workspace_process_context: IWorkspaceProcessContext,
     logger: logging.Logger,
     debug_crash_flags: Optional[Mapping[str, int]] = None,
-) -> Iterable[Optional[SerializableErrorInfo]]:
+) -> Iterable[Optional[PartitionBackfill]]:
     instance = workspace_process_context.instance
     backfills = instance.get_backfills(status=BulkActionStatus.REQUESTED)
```

Comment:
I think one question here is whether we actually need to re-fetch the backfill object itself from the DB every 2 seconds - if this PR caused problems I'd suspect hammering the backfill DB to be the most likely cause. If it doesn't need to fetch the updated backfill on each iteration and just needs to run other checkup logic more frequently, i'd suggest moving all logic that fetches the list of backfills to the outer function and running the refetch less frequently / only after a certain amount of time has passed. If it does need to re-fetch the backfill object on each 2 second tick to work, then it may be unavoidable

Reply:
The main risk would be that someone cancels the backfill, but we go on submitting runs

Reply:
I think in the job backfill case, at least, we already pause every chunk of runs that we schedule to check for updates to the status.  And that would hammer the db every second?